packages/api/src/brevo-contact/brevo-contact-import.service.ts

```diff
@@ -29,114 +46,151 @@ export class BrevoContactImportService {
-                    });
-                    if (success) created++;
-                    else failed++;
-                }
-            } catch (err) {
-                console.error(err);
+            } catch (validationError) {
+                console.error(validationError);
+                failedColumns.push(row);
                 failed++;
             }
         }
-        return { created, updated, failed };
+        if (created + updated + failed === 0) {
+            return { created, updated, failed, failedColumns, errorMessage: "No contacts found." };
+        }
+
+        return { created, updated, failed, failedColumns };
     }
 
-    async parseCsvToBrevoContacts(csvContent: string, redirectUrl: string): Promise<CreateDoubleOptInContactData[]> {
-        const brevoContacts: CreateDoubleOptInContactData[] = [];
-
-        return new Promise((resolve, reject) => {
-            parseString(csvContent, { headers: true, delimiter: ";" })
-                .on("error", (error) => {
-                    console.error(error);
-                    reject(error);
-                })
-                // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                .on("data", async (row: Record<string, any>) => {
-                    try {
-                        const contactData = await this.processRow(row, redirectUrl);
-                        brevoContacts.push(contactData);
-                    } catch (validationError) {
-                        console.error(validationError);
-                        reject(validationError);
-                    }
-                })
-                .on("end", (rowCount: number) => {
-                    console.log(`Parsed ${rowCount} rows`);
-                    resolve(brevoContacts);
+    private async createOrUpdateBrevoContact(
+        contact: CreateDoubleOptInContactData,
+        scope: EmailCampaignScopeInterface,
+        targetGroupBrevoIds: number[],
+    ): Promise<"created" | "updated" | "error"> {
+        try {
+            let brevoContact;
+            try {
+                brevoContact = await this.brevoApiContactsService.findContact(contact.email, scope);
+            } catch (error) {
+                // Brevo throws 404 error if no contact was found
+                if (!isErrorFromBrevo(error)) {
+                    throw error;
+                }
+                if (error.response.statusCode !== 404) {
+                    throw error;
+                }
+            }
+            if (brevoContact && !brevoContact.emailBlacklisted) {
+                const mainTargetGroupForScope = await this.targetGroupsService.createIfNotExistMainTargetGroupForScope(scope);
+
+                const updatedBrevoContact = await this.brevoApiContactsService.updateContact(
+                    brevoContact.id,
+                    { ...contact, listIds: [mainTargetGroupForScope.brevoId, ...targetGroupBrevoIds, ...brevoContact.listIds] },
```

Comment:
The targetGroupBrevoIds are only the manuallyAssigned targetGroupIds if I understand it correctly. Does this also include the contact in the "main" list of the target group? Since the manuallyAssigned group is only a helper target group? Does the contact end up in the recipients this way?

Reply:
I did not specifically check this code because I didn't change anything there. I'll check it now

Reply:
Thank you. This did not work correctly as you said, fixed in: 24a42a930ddf8939027edf0e06151e15aae1ba40

Reply:
72332f98f4f7b1a8b0b8d390fca671caedd55607

added validation for main lists and for the correct scope: 72332f98f4f7b1a8b0b8d390fca671caedd55607

manually assigned target group ids are defined here:
https://github.com/vivid-planet/comet-brevo-module/blob/main/packages/api/src/target-group/entity/target-group-entity.factory.ts#L78

and wouldn't be found when looking for targetGroups by brevoId.

test list ids are not yet merged

Reply:
Thanks for checking! 
Looking good now